nexus/src/app/deployment.rs

```diff
@@ -0,0 +1,344 @@
+            Ok(blueprints.target.clone())
+        } else {
+            Err(Error::not_found_by_id(ResourceType::Blueprint, &new_target_id))
+        }
+    }
+
+    async fn blueprint_planning_context(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<PlanningContext, Error> {
+        let creator = self.id.to_string();
+        let datastore = self.datastore();
+        let collection = datastore
+            .inventory_get_latest_collection(opctx, SQL_LIMIT_INVENTORY)
+            .await?
+            .ok_or_else(|| {
+                Error::unavail("no recent inventory collection available")
+            })?;
+
+        let sled_rows = {
+            let mut all_sleds = Vec::new();
+            let mut paginator = Paginator::new(SQL_BATCH_SIZE);
+            while let Some(p) = paginator.next() {
+                let batch =
+                    datastore.sled_list(opctx, &p.current_pagparams()).await?;
+                paginator =
+                    p.found_batch(&batch, &|s: &nexus_db_model::Sled| s.id());
+                all_sleds.extend(batch);
+            }
+            all_sleds
+        };
+
+        let mut zpools_by_sled_id = {
+            let mut zpools = BTreeMap::new();
+            let mut paginator = Paginator::new(SQL_BATCH_SIZE);
+            while let Some(p) = paginator.next() {
+                let batch = datastore
+                    .zpool_list_all_external(opctx, &p.current_pagparams())
+                    .await?;
+                paginator =
+                    p.found_batch(&batch, &|z: &nexus_db_model::Zpool| z.id());
+                for z in batch {
+                    let sled_zpool_names =
+                        zpools.entry(z.sled_id).or_insert_with(BTreeSet::new);
+                    // It's unfortunate that Nexus knows how Sled Agent
+                    // constructs zpool names, but there's not currently an
+                    // alternative.
+                    let zpool_name_generated =
+                        illumos_utils::zpool::ZpoolName::new_external(z.id())
+                            .to_string();
+                    let zpool_name = ZpoolName::from_str(&zpool_name_generated)
+                        .map_err(|e| {
+                            Error::internal_error(&format!(
+                                "unexpectedly failed to parse generated \
+                                zpool name: {}: {}",
+                                zpool_name_generated, e
+                            ))
+                        })?;
+                    sled_zpool_names.insert(zpool_name);
+                }
+            }
+            zpools
+        };
+
+        let sleds = sled_rows
+            .into_iter()
+            .map(|sled_row| {
+                let sled_id = sled_row.id();
+                let subnet = Ipv6Subnet::<SLED_PREFIX>::new(sled_row.ip());
+                let zpools = zpools_by_sled_id
```

Comment:
I'm on the fence.  Right now that seems likely true.  I'm not sure it always will be.  We may wind up filtering out sleds that are decommissioned.   If we do that, we should probably also exclude their zpools from the query, too.  But that may not be worth the complexity, if decommissioned sleds are pretty uncommon.  Anyway, if somehow this were wrong, I'm not sure we'd want to fail the whole operation.  I could really go either way and I'm usually in favor of extra checks that things aren't off the rails, but my inclination is not to fail in this case because it seems more likely that there's just some runtime condition that we didn't expect that's actually okay than it does that we've got this kind of logic bug.